### DISCUSSION TITLE
Inheriting from ExternalTaskSensor works, but dependencies don't show up in UI

### DISCUSSION THREAD

**Discussion by davidavdav** [apache/airflow#19582] [Q&A] [answered]
Hello, 

We use a cross-dag dependency, but our triggers are not datetime based but event triggered (say, a file upload).  We then use an event-ID as the dag `run_id`, and we can use this across DAGs.  Since events trigger different DAGs, but can be about the same event-ID, it makes sense to have a `ExternalTaskSensor` that just inspects the state of the dag `run_id` as this event-ID in the external DAG. 

By simple inheritance of `ExternalTaskSensor` we can get this working quite easily:
```python
class ExternalDagSensor(ExternalTaskSensor):
    template_fields = ['external_dag_id', 'run_id']
    ui_color = '#e6ccff'
    def __init__(self, *, external_dag_id: str, run_id: str = None, **kwargs):
        super().__init__(external_dag_id=external_dag_id, execution_delta=None, execution_date_fn=None, **kwargs)
        self.run_id = run_id

@provide_session
def poke(self, context, session=None):
    run_id = self.run_id or context['run_id']
    ...
```

The only thing that's missing is that this DAG dependency doesn't show up in the UI under "Browse... DAG Dependencies"

I've experimented with explicitly defining a property `operator_extra_links`, and with making this our own class `MyExternalTaskSensorLink`, but to no avail. 

I have the feeling that we must register our class `ExternalDagSensor` as something that mediates DAGs, but I can't find how.  Maybe we have to register something with `ProvidersManager`?

Thanks!

**Comment by potiuk** [comment, marked as answer]
I am afraid we have no support for this. Detecting dependencies seems to be based on specific operators used:

https://github.com/apache/airflow/blob/main/airflow/serialization/serialized_objects.py#L480

It would be nice feature request or PR to add support for all tasks derived from it, but I think it would not be trivial (and indeed would require some registration). You can also modify the code of your airflow to add your class and double check if it works first.

**Comment by davidavdav** [comment]
Hello, 

Thanks for the pointer.  I see a few solutions.  One is changing the code of airflow (I can submit a PR):
```diff
diff --git a/airflow/serialization/serialized_objects.py b/airflow/serialization/serialized_objects.py
index bdbaea826..78eeff95d 100644
--- a/airflow/serialization/serialized_objects.py
+++ b/airflow/serialization/serialized_objects.py
@@ -48,6 +48,8 @@ from airflow.settings import json
 from airflow.utils.code_utils import get_python_source
 from airflow.utils.module_loading import import_string
 from airflow.utils.task_group import TaskGroup
+from airflow.sensors.external_task import ExternalTaskSensor
+from airflow.operators.trigger_dagrun import TriggerDagRunOperator
 
 try:
     # isort: off
@@ -338,14 +340,14 @@ class DependencyDetector:
     @staticmethod
     def detect_task_dependencies(task: BaseOperator) -> Optional['DagDependency']:
         """Detects dependencies caused by tasks"""
-        if task.task_type == "TriggerDagRunOperator":
+        if isinstance(task, TriggerDagRunOperator):
             return DagDependency(
                 source=task.dag_id,
                 target=getattr(task, "trigger_dag_id"),
                 dependency_type="trigger",
                 dependency_id=task.task_id,
             )
-        elif task.task_type == "ExternalTaskSensor":
+        elif isinstance(task, ExternalTaskSensor):
             return DagDependency(
                 source=getattr(task, "external_dag_id"),
                 target=task.dag_id,
```
Another is to overload an extra property in our own class `ExternalDagSensor` (see above):
```python
class ExternalDagSensor(ExternalTaskSensor):
    task_type = "ExternalTaskSensor"
    ...
```
Both seem to work for us, but I find the latter slightly less elegant, it will probably show the wrong operator in various database entries.

**Reply by potiuk** [reply]
Can you make a PR with the fix (the first case sounds better)

**Reply by davidavdav** [reply]
I referenced this this discussion from PR #19683, but it didn't get picked up by github, hence this backreference
